staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go

```diff
@@ -315,7 +381,124 @@ func (s *APIAggregator) RemoveAPIService(apiServiceName string) {
 	s.GenericAPIServer.Handler.NonGoRestfulMux.Unregister(proxyPath)
 	s.GenericAPIServer.Handler.NonGoRestfulMux.Unregister(proxyPath + "/")
 	delete(s.proxyHandlers, apiServiceName)
+	s.deleteApiSpec(apiServiceName)
+	s.updateOpenAPISpec()
 
 	// TODO unregister group level discovery when there are no more versions for the group
 	// We don't need this right away because the handler properly delegates when no versions are present
 }
+
+func (_ *APIAggregator) loadOpenAPISpec(p *proxyHandler, r *http.Request) (*spec.Swagger, error) {
+	value := p.handlingInfo.Load()
+	if value == nil {
+		return nil, nil
+	}
+	handlingInfo := value.(proxyHandlingInfo)
+	if handlingInfo.local {
+		return nil, nil
+	}
+	loc, err := p.routing.ResolveEndpoint(handlingInfo.serviceNamespace, handlingInfo.serviceName)
+	if err != nil {
+		return nil, fmt.Errorf("missing route")
+	}
+	host := loc.Host
+
+	var w io.Reader
+	req, err := http.NewRequest("GET", "/swagger.json", w)
+	if err != nil {
+		return nil, err
+	}
+	req.URL.Scheme = "https"
+	req.URL.Host = host
+
+	req = req.WithContext(context.Background())
+	// Get user from the original request
+	ctx, ok := p.contextMapper.Get(r)
+	if !ok {
+		return nil, fmt.Errorf("missing context")
+	}
+	user, ok := genericapirequest.UserFrom(ctx)
+	if !ok {
+		return nil, fmt.Errorf("missing user")
+	}
+	proxyRoundTripper := transport.NewAuthProxyRoundTripper(user.GetName(), user.GetGroups(), user.GetExtra(), handlingInfo.proxyRoundTripper)
+	res, err := proxyRoundTripper.RoundTrip(req)
+	if err != nil {
+		return nil, err
+	}
+	if res.StatusCode != http.StatusOK {
+		return nil, errors.New(res.Status)
+	}
+	buf := new(bytes.Buffer)
+	buf.ReadFrom(res.Body)
+	bytes := buf.Bytes()
+	var s spec.Swagger
+	if err := json.Unmarshal(bytes, &s); err != nil {
+		return nil, err
+	}
+	return &s, nil
+}
+
+// Returns true if any Spec is loaded
+func (s *APIAggregator) tryLoadingOpenAPISpecs(r *http.Request) bool {
+	s.specMutex.Lock()
+	defer s.specMutex.Unlock()
+	if len(s.toLoadAPISpec) == 0 {
+		return false
+	}
+	loaded := false
+	newList := map[string]int{}
+	for name, retries := range s.toLoadAPISpec {
+		if retries >= LOAD_OPENAPI_SPEC_MAX_RETRIES {
```

Comment:
shouldn't this be `>`? Otherwise, the first failure will set it to 1 and this if-clause ignores it.